server/src/main/java/org/opensearch/indices/IndicesService.java

```diff
@@ -1754,6 +1755,9 @@ public boolean canCache(ShardSearchRequest request, SearchContext context) {
         if (context.getQueryShardContext().isCacheable() == false) {
             return false;
         }
+        if (!(context.searcher().getIndexReader().getReaderCacheHelper() instanceof OpenSearchDirectoryReader.DelegatingCacheHelper)) {
```

Comment:
Looks it should be `context.searcher().getDirectoryReader()` ?
Can we add a UT around this?